src/lib/storage/dictionary_segment.cpp

```diff
@@ -122,6 +123,207 @@ const ValueID DictionarySegment<T>::null_value_id() const {
+
+  ChunkOffset res = 0;
+
+  resolve_compressed_vector_type(*_attribute_vector, [&](const auto& attribute_vector) {
+    if (_sort_order.value() == OrderByMode::Ascending || _sort_order.value() == OrderByMode::AscendingNullsLast) {
+      if (position_filter != nullptr) {
+        // TODO(someone): if possible remove dictionary lookup in lambda function and just work on the value ids
+        const auto result =
+            std::lower_bound(position_filter->cbegin(), position_filter->cend(), casted_search_value,
+                             [&](const auto& row_id, const auto& search) {
+                               return _dictionary->operator[](_decompressor->get(row_id.chunk_offset)) < search;
+                             });
+        if (result == position_filter->cend()) {
+          res = INVALID_CHUNK_OFFSET;
+        } else {
+          res = static_cast<ChunkOffset>(std::distance(position_filter->cbegin(), result));
+        }
+      } else {
+        const auto begin = attribute_vector.cbegin() + non_null_begin;
+        const auto end = attribute_vector.cbegin() + non_null_end;
+        const auto result = std::lower_bound(
+            begin, end, casted_search_value,
+            [&](const auto& value_id, const auto& search) { return _dictionary->operator[](value_id) < search; });
+        if (result == end) {
+          res = INVALID_CHUNK_OFFSET;
+        } else {
+          res = static_cast<ChunkOffset>(std::distance(attribute_vector.cbegin(), result));
+        }
+      }
+    } else {
+      if (position_filter != nullptr) {
+        // TODO(someone): if possible remove dictionary lookup in lambda function and just work on the value ids
+        const auto result =
+            std::lower_bound(position_filter->cbegin(), position_filter->cend(), casted_search_value,
+                             [&](const auto& row_id, const auto& search) {
+                               return _dictionary->operator[](_decompressor->get(row_id.chunk_offset)) > search;
+                             });
+        if (result == position_filter->cend()) {
+          res = INVALID_CHUNK_OFFSET;
+        } else {
+          res = static_cast<ChunkOffset>(std::distance(position_filter->cbegin(), result));
+        }
+      } else {
+        const auto begin = attribute_vector.cbegin() + non_null_begin;
+        const auto end = attribute_vector.cbegin() + non_null_end;
+        const auto result = std::lower_bound(
+            begin, end, casted_search_value,
+            [&](const auto& value_id, const auto& search) { return _dictionary->operator[](value_id) > search; });
+        if (result == end) {
+          res = INVALID_CHUNK_OFFSET;
+        } else {
+          res = static_cast<ChunkOffset>(std::distance(attribute_vector.cbegin(), result));
+        }
+      }
+    }
+  });
+
+  return res;
+}
+
+template <typename T>
+ChunkOffset DictionarySegment<T>::get_last_offset(const AllTypeVariant& search_value,
```

Comment:
This function is a duplication of `get_non_null_end_offset` except the calculation of `const auto result`. Additionally within this function you created code duplication as mentioned [before](https://github.com/hyrise/hyrise/pull/1394/files#r257996161).
You could reduce the code by adding case distinction for the calculation of the `result` value.